README.md
Russell King's irqbalanced fork for ARM devices
===============================================

This is a fork of irqbalanced to add proper support for ARM devices.
The regular irqbalanced contains several IRQ misclassifications, which renders it useless
on these systems. Also, the way the number of CPUs were being determined was faulty.

The sources are hosted here with Russell King's permission, to make it easier to build irqbalanced with Yocto, for example.

Building follows the regular autotools pattern, which is:

    ./autogen.sh
    ./configure
    make
    make install


debian/changelog
irqbalance-imx (0.56.99.2) unstable; urgency=medium

  * resolve conflicts with Debian irqbalance package

 -- Josua Mayer <privacy@not.given>  Tue, 14 Mar 2017 13:34:11 +0000

irqbalance-imx (0.56.99.1) unstable; urgency=low

  * rewrote rules file
  * added init script

 -- Josua Mayer <privacy@not.given>  Sat, 22 Nov 2014 18:36:29 +0100

irqbalance-imx (0.56.99.0) unstable; urgency=low

  * Added -imx suffix to distringuish from upstream

 -- Josua Mayer <privacy@not.given>  Mon, 03 Nov 2014 15:10:52 +0100

irqbalance (0.56-1ubuntu4-rmk1) precise; urgency=low

  * Fix CPU topology parsing in the presence of cpufreq
  * Add interrupt classifications for i.MX6 hardware

 -- Russell King <rmk@arm.linux.org.uk>  Wed, 05 Feb 2014 17:22:03 +0000

irqbalance (0.56-1ubuntu4) precise; urgency=low

  * Apply patch from Stephen Gran to fix a SEGV on some systems.

 -- Adam Conrad <adconrad@ubuntu.com>  Fri, 03 Feb 2012 22:13:07 -0700

irqbalance (0.56-1ubuntu3) natty; urgency=low

  * Fixed debian/irqbalance.config to correctly pickup existing
    settings from /etc/default/irqbalance (LP: #705429)

 -- James Page <james.page@canonical.com>  Thu, 27 Jan 2011 15:36:50 +0000

irqbalance (0.56-1ubuntu2) natty; urgency=low

  * Build-depend on quilt.

 -- Matthias Klose <doko@ubuntu.com>  Thu, 20 Jan 2011 01:37:32 +0100

irqbalance (0.56-1ubuntu1) natty; urgency=low

  * Merge from debian unstable.  Remaining changes:
    - Update maintainer according to spec.
    - debian/rules, debian/irqbalance.dirs, debian/irqbalance.upstart:
      Add upstart support.
    - debian/control: Add libcap-ng support.
    - debian/control: Update description to give the user a bit more
      information on the package.
    - debian/irqbalance.{config,postinst}: Ask the user ift hey want
      to disable irqbalance or not. (LP: #573657)

 -- Chuck Short <zulcss@ubuntu.com>  Mon, 10 Jan 2011 10:41:18 -0400

irqbalance (0.56-1) unstable; urgency=low

  * New upstream 
  * 01-cputree-parse.patch was dropped by redhat
  * Update debian/watch
  * Fix out-of-date-standards-version

 -- Anibal Monsalve Salazar <anibal@debian.org>  Fri, 09 Jul 2010 15:20:02 +1000

irqbalance (0.56-0ubuntu2) maverick; urgency=low

  * debian/irqbalance.{config,postinst}: Ask the user if they want to 
    disable irqbalance or not. (LP: #573657)

 -- Chuck Short <zulcss@ubuntu.com>  Wed, 08 Sep 2010 09:01:22 -0400

irqbalance (0.56-0ubuntu1) maverick; urgency=low

  * New upstream release, this release offers better support for 2.3.35.
  * debian/irqbalance.upstart: Dont respawn. (LP: #594509)
  * debian/patches/do-irqbalance-pid.patch: Dropped no longer needed.

 -- Chuck Short <zulcss@ubuntu.com>  Wed, 23 Jun 2010 10:39:31 -0400

irqbalance (0.55+20091017-5) unstable; urgency=low

  * Fix incorrect apostrophe in IRQ's string
    Patch by Brian Murray
    Closes: 571160
  * Debconf translations:
    Vietnamese: Clytie Siddall. Closes: 576129

 -- Anibal Monsalve Salazar <anibal@debian.org>  Sun, 11 Apr 2010 08:07:38 +1000

irqbalance (0.55+20091017-4) unstable; urgency=low

  * Debian source format is 3.0 (quilt)
  * Fix out-of-date-standards-version
  * Update description
    Closes: 529878

 -- Anibal Monsalve Salazar <anibal@debian.org>  Fri, 29 Jan 2010 17:40:59 +1100

irqbalance (0.55+20091017-3.1) unstable; urgency=low

  * Non-maintainer upload.
  * Fix broken Italian translation file that is b0rking automated
    processes on i18n.debian.net. Closes: #556110

 -- Christian Perrier <bubulle@debian.org>  Sun, 24 Jan 2010 15:27:07 +0100

irqbalance (0.55+20091017-3ubuntu3) maverick; urgency=low

  * debian/patches/do-irqbalance-pid.patch: Upstart needs a pid so it can shut down.
    (LP: #580674)
  * debian/irqbalance.upstart: Update upstart script. 

 -- Chuck Short <zulcss@ubuntu.com>  Tue, 25 May 2010 17:38:47 -0400

irqbalance (0.55+20091017-3ubuntu2) lucid; urgency=low

  * Cleanup some grammatical errors (Closes: 571160)

 -- Brian Murray <brian@ubuntu.com>  Tue, 23 Feb 2010 14:31:22 -0800

irqbalance (0.55+20091017-3ubuntu1) lucid; urgency=low

  * Merge from debian testing.  Remaining changes:
    - Update maintainer according to Spec.
    - debian/rules, debian/irqbalance.dirs, debian/irqbalance.upstart:
      Add upstart support.
    - debian/control: Add libcab-ng support.
    - debian/control: Update description to give the user a bit more information 
      on the package.

 -- Chuck Short <zulcss@ubuntu.com>  Mon, 04 Jan 2010 15:11:10 +0000

irqbalance (0.55+20091017-3) unstable; urgency=low

  * Debconf translations:
    Japanese. Closes: 552048
    Italian. Closes: 556110

 -- Anibal Monsalve Salazar <anibal@debian.org>  Mon, 23 Nov 2009 14:21:35 +1100

irqbalance (0.55+20091017-2ubuntu3) lucid; urgency=low

  * debian/irqbalance.upstart: Fix irqbalance upstart script. (LP: #499864) 

 -- Chuck Short <zulcss@ubuntu.com>  Sat, 26 Dec 2009 22:12:26 -0500

irqbalance (0.55+20091017-2ubuntu2) lucid; urgency=low

  * debian/control: Add libcap-ng support. 
  * debian/irqbalance.init, debian/irqbalance.upstart, 
    debian/rules, debian/irqbalance.dirs: Add upstart support.

 -- Chuck Short <zulcss@ubuntu.com>  Mon, 14 Dec 2009 09:47:39 -0500

irqbalance (0.55+20091017-2ubuntu1) lucid; urgency=low

  * debian/control:
    - Update maintainer according to spec.
    - Added note about irqbalance and how it is not useful if you dont have
      more than one CPU or CPU core.
    - Added hardening wrapper. 

 -- Chuck Short <zulcss@ubuntu.com>  Thu, 26 Nov 2009 11:26:06 -0500

irqbalance (0.55+20091017-2) unstable; urgency=low

  * Debconf translations:
    Russian. Closes: 551502
    French. Closes: 551585
    Italian
    Basque. Closes: 552073
    Finnish. Closes: 552155
    Czech. Closes: 552451
    Swedish. Closes: 552735
    Portuguese. Closes: 553082
    German. Closes: 553394
    Spanish. Closes: 553912
    Galician. Closes: 554632

 -- Anibal Monsalve Salazar <anibal@debian.org>  Sat, 07 Nov 2009 20:56:54 +1100


types.h
#ifndef _INCLUDE_GUARD_TYPES_H
#define _INCLUDE_GUARD_TYPES_H

#include <glib.h>

#include "cpumask.h"

#define	BALANCE_NONE		0
#define BALANCE_PACKAGE 	1
#define BALANCE_CACHE		2
#define BALANCE_CORE		3

#define IRQ_OTHER       0
#define IRQ_LEGACY      1
#define IRQ_SCSI        2
#define IRQ_TIMER       3
#define IRQ_ETH         4
#define IRQ_GETH	5
#define IRQ_TGETH	6
#define IRQ_AV		7


struct package {
	uint64_t	workload;
	int	number;

	cpumask_t	mask;

	int class_count[8];

	GList	*cache_domains;
	GList 	*interrupts;
};

struct cache_domain {
	uint64_t	workload;
	int	number;

	int marker;

	cpumask_t	mask;

	cpumask_t	package_mask;

	int class_count[8];

	GList	*cpu_cores;
	GList 	*interrupts;
};


struct cpu_core {
	uint64_t	workload;
	int	number;

	int	marker;

	int class_count[8];

	cpumask_t	package_mask;
	cpumask_t	cache_mask;
	cpumask_t	mask;

	GList 	*interrupts;
};

struct interrupt {
	uint64_t	workload;

	int	balance_level;

	int	number;
	int	class;

	uint64_t	count;
	uint64_t	old_count;
	uint64_t	extra;

	cpumask_t	mask;
	cpumask_t	old_mask;
	

	cpumask_t	numa_mask;
	cpumask_t	allowed_mask;

	/* user/driver provided for smarter balancing */
	cpumask_t	node_mask;
};


#endif


irqlist.c
/* 
 * Copyright (C) 2006, Intel Corporation
 * 
 * This file is part of irqbalance
 *
 * This program file is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; version 2 of the License.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program in a file named COPYING; if not, write to the 
 * Free Software Foundation, Inc., 
 * 51 Franklin Street, Fifth Floor, 
 * Boston, MA 02110-1301 USA
 */

/*
 * This file has the basic functions to manipulate interrupt metadata
 */
#include "config.h"
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <dirent.h>

#include "types.h"
#include "irqbalance.h"

GList *interrupts;



void get_affinity_hint(struct interrupt *irq, int number)
{
	char buf[PATH_MAX];
	char *line = NULL;
	size_t size = 0;
	FILE *file;
	sprintf(buf, "/proc/irq/%i/affinity_hint", number);
	file = fopen(buf, "r");
	if (!file)
		return;
	if (getline(&line, &size, file)==0) {
		free(line);
		fclose(file);
		return;
	}
	cpumask_parse_user(line, strlen(line), irq->node_mask);
	fclose(file);
	free(line);
}

/*
 * This function classifies and reads various things from /proc about a specific irq 
 */
static void investigate(struct interrupt *irq, int number)
{
	DIR *dir;
	struct dirent *entry;
	char *c, *c2;
	int nr , count = 0;
	char buf[PATH_MAX];
	sprintf(buf, "/proc/irq/%i", number);
	if ((dir = opendir(buf)) == NULL) {
		return;
	}
	do {
		entry = readdir(dir);
		if (!entry)
			break;
		if (strcmp(entry->d_name,"smp_affinity")==0) {
			char *line = NULL;
			size_t size = 0;
			FILE *file;
			sprintf(buf, "/proc/irq/%i/smp_affinity", number);
			file = fopen(buf, "r");
			if (!file)
				continue;
			if (getline(&line, &size, file)==0) {
				free(line);
				fclose(file);
				continue;
			}
			cpumask_parse_user(line, strlen(line), irq->mask);
			fclose(file);
			free(line);
		} else if (strcmp(entry->d_name,"allowed_affinity")==0) {
			char *line = NULL;
			size_t size = 0;
			FILE *file;
			sprintf(buf, "/proc/irq/%i/allowed_affinity", number);
			file = fopen(buf, "r");
			if (!file)
				continue;
			if (getline(&line, &size, file)==0) {
				free(line);
				fclose(file);
				continue;
			}
			cpumask_parse_user(line, strlen(line), irq->allowed_mask);
			fclose(file);
			free(line);
		} else if (strcmp(entry->d_name,"affinity_hint")==0) {
			get_affinity_hint(irq, number);
		} else {
			irq->class = find_class(irq, entry->d_name);
		}

	} while (entry);
	closedir(dir);	
	irq->balance_level = map_class_to_level[irq->class];

	for (nr = 0; nr < NR_CPUS; nr++)
		if (cpu_isset(nr, irq->allowed_mask))
			count++;

	/* if there is no choice in the allowed mask, don't bother to balance */
	if (count<2)
		 irq->balance_level = BALANCE_NONE;
		

	/* next, check the IRQBALANCE_BANNED_INTERRUPTS env variable for blacklisted irqs */
	c = c2 = getenv("IRQBALANCE_BANNED_INTERRUPTS");
	if (!c)
		return;

	do {
		c = c2;
		nr = strtoul(c, &c2, 10);
		if (c!=c2 && nr == number)
			irq->balance_level = BALANCE_NONE;
	} while (c!=c2 && c2!=NULL);
}


/*
 * Set the number of interrupts received for a specific irq;
 * create the irq metadata if there is none yet
 */
void set_interrupt_count(int number, uint64_t count)
{
	GList *item;
	struct interrupt *irq;

	if (count < MIN_IRQ_COUNT && !one_shot_mode)
		return; /* no need to track or set interrupts sources without any activity since boot
		 	   but allow for a few (20) boot-time-only interrupts */

	item = g_list_first(interrupts);
	while (item) {
		irq = item->data;

		if (irq->number == number) {
			irq->count = count;
			/* see if affinity_hint changed */
			get_affinity_hint(irq, number);
			return;
		}
		item = g_list_next(item);
	}
	/* new interrupt */
	irq = malloc(sizeof(struct interrupt));
	if (!irq)
		return;
	memset(irq, 0, sizeof(struct interrupt));
	irq->number = number;
	irq->count = count;
	irq->allowed_mask = CPU_MASK_ALL;
	investigate(irq, number);
	interrupts = g_list_append(interrupts, irq);
}

/* 
 * Add extra irqs to a specific irq metadata structure;
 * if no such metadata exists, do nothing at all
 */
void add_interrupt_count(int number, uint64_t count, int type)
{
	GList *item;
	struct interrupt *irq;

	if (!count)
		return;

	item = g_list_first(interrupts);
	while (item) {
		irq = item->data;
		item = g_list_next(item);

		if (irq->number == number) {
			irq->extra += count;
			if (irq->class < type && irq->balance_level != BALANCE_NONE)  {
				irq->class = type;
				irq->balance_level = map_class_to_level[irq->class];
			}
			return;
		}
	}
}

/*
 * Set the numa affinity mask for a specific interrupt if there
 * is metadata for the interrupt; do nothing if no such data
 * exists.
 */
void add_interrupt_numa(int number, cpumask_t mask, int type)
{
	GList *item;
	struct interrupt *irq;

	item = g_list_first(interrupts);
	while (item) {
		irq = item->data;
		item = g_list_next(item);

		if (irq->number == number) {
			cpus_or(irq->numa_mask, irq->numa_mask, mask);
			if (irq->class < type && irq->balance_level != BALANCE_NONE)  {
				irq->class = type;
				irq->balance_level = map_class_to_level[irq->class];
			}
			return;
		}
	}
}

void calculate_workload(void)
{
	int i;
	GList *item;
	struct interrupt *irq;

	for (i=0; i<7; i++)
		class_counts[i]=0;
	item = g_list_first(interrupts);
	while (item) {
		irq = item->data;
		item = g_list_next(item);

		irq->workload = irq->count - irq->old_count + irq->workload/3 + irq->extra;
		class_counts[irq->class]++;
		irq->old_count = irq->count;
		irq->extra = 0;
	}
}

void reset_counts(void)
{
	GList *item;
	struct interrupt *irq;
	item = g_list_first(interrupts);
	while (item) {
		irq = item->data;
		item = g_list_next(item);
		irq->old_count = irq->count;
		irq->extra = 0;

	}
}

void dump_workloads(void)
{
	GList *item;
	struct interrupt *irq;
	item = g_list_first(interrupts);
	while (item) {
		irq = item->data;
		item = g_list_next(item);

		printf("Interrupt %i (class %s) has workload %lu \n", irq->number, classes[irq->class], (unsigned long)irq->workload);

	}
}


static gint sort_irqs(gconstpointer A, gconstpointer B)
{
	struct interrupt *a, *b;
	a = (struct interrupt*)A;
	b = (struct interrupt*)B;

	if (a->class < b->class)
		return 1;
	if (a->class > b->class)
		return -1;
	if (a->workload < b->workload)
		return 1;
	if (a->workload > b->workload)
		return -1;
	if (a<b)
		return 1;
	return -1;
	
}

void sort_irq_list(void)
{
	/* sort by class first (high->low) and then by workload (high->low) */
	interrupts = g_list_sort(interrupts, sort_irqs);
}


classify.c
#include "config.h"
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>

#include "irqbalance.h"
#include "types.h"


char *classes[] = {
	"other",
	"legacy",
	"storage",
	"timer",
	"ethernet",
	"gbit-ethernet",
	"10gbit-ethernet",
	"av",
	0
};

int map_class_to_level[8] = 
{ BALANCE_PACKAGE, BALANCE_CACHE, BALANCE_CACHE, BALANCE_NONE, BALANCE_CORE, BALANCE_CORE, BALANCE_CORE, BALANCE_CORE };


int class_counts[8];

/*

NOTE NOTE although that this file has a hard-coded list of modules, something missing is not
          a big deal; the types are also set based on PCI class information when available.

*/

/*

   Based on the original irqbalance code which is:

   Copyright (C) 2003 Red Hat, Inc. All rights reserved.
                                                                                                             
   Usage and distribution of this file are subject to the Gnu General Public License Version 2 
   that can be found at http://www.gnu.org/licenses/gpl.txt and the COPYING file as
   distributed together with this file is included herein by reference.  
                                                                                                                      
   Author: Arjan van de Ven   <arjanv@redhat.com>

*/

static char *legacy_modules[] = {
 	"PS/2",
 	"serial",
 	"i8042",
 	"acpi",
	"floppy",
	"parport",
 	"keyboard",
 	"usb-ohci",
	"usb-uhci",
	"uhci_hcd",
	"ohci_hcd",
	"ehci_hcd",
	"EMU10K1",
	"ci_hdrc",
	0
};

static char *timer_modules[] = {
 	"rtc",
 	"timer",
 	"twd",
	0
};

static char *storage_modules[] = {
	"aic7xxx",
	"aic79xx",
	"ide",
	"cciss",
	"cpqarray",
	"qla2",
	"megaraid",
	"fusion",
	"libata",
	"ohci1394",
	"sym53c8xx",
	"ahci",
	"mmc",
	0
};

static char *ethernet_modules[] = {
	"eth",
	"e100",
	"eepro100",
	"orinoco_cs",
	"wvlan_cs",
	"3c5",	
	"HiSax",
	"skge",
	"sky2",
	0
};

static char *av_modules[] = {
	"VPU",
	"ipu",
	"galcore",
	"hdmi",
	"spdif",
	0
};


int find_class(struct interrupt *irq, char *moduletext)
{
	int guess = IRQ_OTHER;
	int i;
	
	if (moduletext == NULL)
		return guess;
		
	for (i=0; legacy_modules[i]; i++)
		if (strstr(moduletext, legacy_modules[i]))
			guess = IRQ_LEGACY;

	for (i=0; storage_modules[i]; i++)
		if (strstr(moduletext, storage_modules[i]))
			guess = IRQ_SCSI;

	for (i=0; timer_modules[i]; i++)
		if (strstr(moduletext, timer_modules[i]))
			guess = IRQ_TIMER;
			
	for (i=0; ethernet_modules[i]; i++)
		if (strstr(moduletext, ethernet_modules[i])) {
			guess = IRQ_ETH;
			if (strstr(moduletext, "-rx"))
				guess = IRQ_GETH;
			if (strstr(moduletext, "-tx"))
				guess = IRQ_TGETH;
		}

	for (i=0; av_modules[i]; i++)
		if (strstr(moduletext, av_modules[i]))
			guess = IRQ_AV;

	if (guess == IRQ_OTHER && irq->number==0)
		guess = IRQ_TIMER;
	
	if (guess >  irq->class)	
		return guess;
        return irq->class;
}


irqbalance.h
#ifndef __INCLUDE_GUARD_IRQBALANCE_H_
#define __INCLUDE_GUARD_IRQBALANCE_H_


#include "constants.h"

#include "cpumask.h"

#include <stdint.h>
#include <glib.h>

#include "types.h"

struct interrupt;

extern int package_count;
extern int cache_domain_count;
extern int core_count;
extern char *classes[];
extern int map_class_to_level[8];
extern int class_counts[8];
extern int debug_mode;
extern int power_mode;
extern int need_cpu_rescan;
extern int one_shot_mode;
extern GList *interrupts;


extern void parse_cpu_tree(void);
extern void clear_work_stats(void);
extern void parse_proc_interrupts(void);
extern void set_interrupt_count(int number, uint64_t count);
extern void add_interrupt_count(int number, uint64_t count, int type);
extern int find_class(struct interrupt *irq, char *string);
extern void add_interrupt_numa(int number, cpumask_t mask, int type);

void calculate_workload(void);
void reset_counts(void);
void dump_workloads(void);
void sort_irq_list(void);
void calculate_placement(void);
void dump_tree(void);

void activate_mapping(void);
void account_for_nic_stats(void);
void check_power_mode(void);
void clear_cpu_tree(void);
void pci_numa_scan(void);

#endif


cputree.c
/* 
 * Copyright (C) 2006, Intel Corporation
 * 
 * This file is part of irqbalance
 *
 * This program file is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; version 2 of the License.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program in a file named COPYING; if not, write to the 
 * Free Software Foundation, Inc., 
 * 51 Franklin Street, Fifth Floor, 
 * Boston, MA 02110-1301 USA
 */

/*
 * This file contains the code to construct and manipulate a hierarchy of processors,
 * cache domains and processor cores.
 */

#include "config.h"
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <dirent.h>

#include <glib.h>

#include "irqbalance.h"


GList *cpus;
GList *cache_domains;
GList *packages;

int package_count;
int cache_domain_count;
int core_count;

/* Users want to be able to keep interrupts away from some cpus; store these in a cpumask_t */
cpumask_t banned_cpus;


/* 
   it's convenient to have the complement of banned_cpus available so that 
   the AND operator can be used to mask out unwanted cpus
*/
static cpumask_t unbanned_cpus;

static void fill_packages(void)
{
	GList *entry;

	entry = g_list_first(cache_domains);
	while (entry) {
		struct package *package;
		struct cache_domain *cache = NULL;
		GList *entry2;

		cache = entry->data;
		entry2 = entry;
		entry = g_list_next(entry);
		if (cache->marker) 
			continue;
		package = malloc(sizeof(struct package));
		if (!package)
			break;
		memset(package, 0, sizeof(struct package));
		package->mask = cache->package_mask;
		package->number = cache->number;
		while (entry2) {
			struct cache_domain *cache2;
			cache2 = entry2->data;
			if (cpus_equal(cache->package_mask, cache2->package_mask)) {
				cache2->marker = 1;
				package->cache_domains = g_list_append(package->cache_domains, cache2);
				if (package->number > cache2->number)
					package->number = cache2->number;
			}
			entry2 = g_list_next(entry2);
		}
		packages = g_list_append(packages, package);
		package_count++;
	}
}

static void fill_cache_domain(void)
{
	GList *entry;

	entry = g_list_first(cpus);
	while (entry) {
		struct cache_domain *cache = NULL;
		struct cpu_core *cpu;
		GList *entry2;
		cpu = entry->data;
		entry2 = entry;
		entry = g_list_next(entry);
		if (cpu->marker) 
			continue;
		cache = malloc(sizeof(struct cache_domain));
		if (!cache)
			break;
		memset(cache, 0, sizeof(struct cache_domain));
		cache->mask = cpu->cache_mask;
		cache->package_mask = cpu->package_mask;
		cache->number = cpu->number;
		cache_domains = g_list_append(cache_domains, cache);
		cache_domain_count++;
		while (entry2) {
			struct cpu_core *cpu2;
			cpu2 = entry2->data;
			if (cpus_equal(cpu->cache_mask, cpu2->cache_mask) && 
			    cpus_equal(cpu->package_mask, cpu2->package_mask)) {
				cpu2->marker = 1;
				cache->cpu_cores = g_list_append(cache->cpu_cores, cpu2);
				if (cpu2->number < cache->number)
					cache->number = cpu2->number;
			}
			entry2 = g_list_next(entry2);
		}
	}
}


static void do_one_cpu(char *path)
{
	struct cpu_core *cpu;
	FILE *file;
	char new_path[PATH_MAX];

	/* skip offline cpus */
	snprintf(new_path, PATH_MAX, "%s/online", path);
	file = fopen(new_path, "r");
	if (file) {
		char *line = NULL;
		size_t size = 0;
		if (getline(&line, &size, file)==0)
			return;
		fclose(file);
		if (line && line[0]=='0') {
			free(line);
			return;
		}
		free(line);
	}

	cpu = malloc(sizeof(struct cpu_core));
	if (!cpu)
		return;
	memset(cpu, 0, sizeof(struct cpu_core));

	cpu->number = strtoul(&path[27], NULL, 10);
	
	cpu_set(cpu->number, cpu->mask);

	/* if the cpu is on the banned list, just don't add it */
	if (cpus_intersects(cpu->mask, banned_cpus)) {
		free(cpu);
		/* even though we don't use the cpu we do need to count it */
		core_count++;
		return;
	}


	/* try to read the package mask; if it doesn't exist assume solitary */
	snprintf(new_path, PATH_MAX, "%s/topology/core_siblings", path);
	file = fopen(new_path, "r");
	cpu_set(cpu->number, cpu->package_mask);
	if (file) {
		char *line = NULL;
		size_t size = 0;
		if (getline(&line, &size, file)) 
			cpumask_parse_user(line, strlen(line), cpu->package_mask);
		fclose(file);
		free(line);
	}

	/* try to read the cache mask; if it doesn't exist assume solitary */
	/* We want the deepest cache level available so try index1 first, then index2 */
	cpu_set(cpu->number, cpu->cache_mask);
	snprintf(new_path, PATH_MAX, "%s/cache/index1/shared_cpu_map", path);
	file = fopen(new_path, "r");
	if (file) {
		char *line = NULL;
		size_t size = 0;
		if (getline(&line, &size, file)) 
			cpumask_parse_user(line, strlen(line), cpu->cache_mask);
		fclose(file);
		free(line);
	}
	snprintf(new_path, PATH_MAX, "%s/cache/index2/shared_cpu_map", path);
	file = fopen(new_path, "r");
	if (file) {
		char *line = NULL;
		size_t size = 0;
		if (getline(&line, &size, file)) 
			cpumask_parse_user(line, strlen(line), cpu->cache_mask);
		fclose(file);
		free(line);
	}

	/* 
	   blank out the banned cpus from the various masks so that interrupts
	   will never be told to go there
	 */
	cpus_and(cpu->cache_mask, cpu->cache_mask, unbanned_cpus);
	cpus_and(cpu->package_mask, cpu->package_mask, unbanned_cpus);
	cpus_and(cpu->mask, cpu->mask, unbanned_cpus);

	cpus = g_list_append(cpus, cpu);
	core_count++;
}

static void dump_irqs(int spaces, GList *interrupts)
{
	struct interrupt *irq;
	while (interrupts) {
		int i;
		for (i=0; i<spaces;i++) printf(" ");
		irq = interrupts->data;
		printf("Interrupt %i (%s/%u) \n", irq->number, classes[irq->class], (unsigned int)irq->workload);
		interrupts = g_list_next(interrupts);
	}
}

void dump_tree(void)
{
	GList *p_iter, *c_iter, *cp_iter;
	struct package *package;
	struct cache_domain *cache_domain;
	struct cpu_core *cpu;

	char buffer[4096];
	p_iter = g_list_first(packages);
	while (p_iter) {
		package = p_iter->data;
		cpumask_scnprintf(buffer, 4096, package->mask);
		printf("Package %i:  cpu mask is %s (workload %lu)\n", package->number, buffer, (unsigned long)package->workload);
		c_iter = g_list_first(package->cache_domains);
		while (c_iter) {
			cache_domain = c_iter->data;
			c_iter = g_list_next(c_iter);
			cpumask_scnprintf(buffer, 4095, cache_domain->mask);
			printf("        Cache domain %i: cpu mask is %s  (workload %lu) \n", cache_domain->number, buffer, (unsigned long)cache_domain->workload);
			cp_iter = cache_domain->cpu_cores;
			while (cp_iter) {
				cpu = cp_iter->data;
				cp_iter = g_list_next(cp_iter);
				printf("                CPU number %i  (workload %lu)\n", cpu->number, (unsigned long)cpu->workload);
				dump_irqs(18, cpu->interrupts);
			}
			dump_irqs(10, cache_domain->interrupts);
		}
		dump_irqs(2, package->interrupts);
		p_iter = g_list_next(p_iter);
	}
}

/*
 * this function removes previous state from the cpu tree, such as
 * which level does how much work and the actual lists of interrupts 
 * assigned to each component
 */
void clear_work_stats(void)
{
	GList *p_iter, *c_iter, *cp_iter;
	struct package *package;
	struct cache_domain *cache_domain;
	struct cpu_core *cpu;

	p_iter = g_list_first(packages);
	while (p_iter) {
		package = p_iter->data;
		package->workload = 0;
		g_list_free(package->interrupts);
		package->interrupts = NULL;
		c_iter = g_list_first(package->cache_domains);
		memset(package->class_count, 0, sizeof(package->class_count));
		while (c_iter) {
			cache_domain = c_iter->data;
			c_iter = g_list_next(c_iter);
			cache_domain->workload = 0;
			cp_iter = cache_domain->cpu_cores;
			g_list_free(cache_domain->interrupts);
			cache_domain->interrupts = NULL;
			memset(cache_domain->class_count, 0, sizeof(cache_domain->class_count));
			while (cp_iter) {
				cpu = cp_iter->data;
				cp_iter = g_list_next(cp_iter);
				cpu->workload = 0;
				g_list_free(cpu->interrupts);
				cpu->interrupts = NULL;
				memset(cpu->class_count, 0, sizeof(cpu->class_count));
			}
		}
		p_iter = g_list_next(p_iter);
	}
}


/*
 * Read from /sys/bus/cpu/devices in preference, as this does not
 * contain other noise which may upset the parsing code (such as
 * /sys/devices/system/cpu/cpufreq).
 */
static const char *cpu_devices_paths[] = {
	"/sys/bus/cpu/devices",
	"/sys/devices/system/cpu",
	NULL,
};

void parse_cpu_tree(void)
{
	static const char *cpu_devices_path;
	DIR *dir;
	struct dirent *entry;
	int i;

	cpus_complement(unbanned_cpus, banned_cpus);

	if (!cpu_devices_path) {
		for (i = 0; cpu_devices_paths[i]; i++) {
			dir = opendir(cpu_devices_paths[i]);
			if (dir) {
				cpu_devices_path = cpu_devices_paths[i];
				break;
			}
		}
		if (!cpu_devices_path)
			return;
	} else {	
		dir = opendir(cpu_devices_path);
		if (!dir)
			return;
	}	

	do {
		entry = readdir(dir);
                if (entry && strlen(entry->d_name)>3 && strstr(entry->d_name,"cpu")) {
			char new_path[PATH_MAX];
			sprintf(new_path, "/sys/devices/system/cpu/%s", entry->d_name);
			do_one_cpu(new_path);
		}
	} while (entry);
	closedir(dir);  

	fill_cache_domain();
	fill_packages();

	if (debug_mode)
		dump_tree();

}


/*
 * This function frees all memory related to a cpu tree so that a new tree
 * can be read
 */
void clear_cpu_tree(void)
{
	GList *item;
	struct cpu_core *cpu;
	struct cache_domain *cache_domain;
	struct package *package;

	while (packages) {
		item = g_list_first(packages);
		package = item->data;
		g_list_free(package->cache_domains);
		g_list_free(package->interrupts);
		free(package);
		packages = g_list_delete_link(packages, item);
	}
	package_count = 0;

	while (cache_domains) {
		item = g_list_first(cache_domains);
		cache_domain = item->data;
		g_list_free(cache_domain->cpu_cores);
		g_list_free(cache_domain->interrupts);
		free(cache_domain);
		cache_domains = g_list_delete_link(cache_domains, item);
	}
	cache_domain_count = 0;


	while (cpus) {
		item = g_list_first(cpus);
		cpu = item->data;
		g_list_free(cpu->interrupts);
		free(cpu);
		cpus = g_list_delete_link(cpus, item);
	}
	core_count = 0;

}
